### PR TITLE
Ray searcher support numpy array input

### PR DESCRIPTION
**Pull request by TheaperDeng**
Related to PR #3045 
Related to issue #3026 

> data should be a dict.
> For data frame, it should be something like {"df": df, "feature_cols": [], "target_col": "value"} or {"df": df, "dt_col": "datatime", "target_cols": ["value"], "feature_cols": []};
> For ndarray, it should be {"x": array1, "y": array2}

- [x] Support dataframe input in new interface
- [x] Change related code to support the dataframe input in new interface
- [x] Support numpy input in new interface
- [x] Change related code to support the numpy input in new interface
- [x] Bug fix to save interface
- [x] UT for the new interface

Current UT for Automl pass on local dev server.
Document for the new interface might be left for next PR since the PR is too large.

### CODE REVIEW

File: pyzoo/zoo/automl/search/ray_tune_search_engine.py [hunk 1]
```diff
@@ -273,51 +317,73 @@ def _prepare_train_func(input_df,
         :param metric: the rewarding metric
         :return: the train function
         """
-        input_df_id = ray.put(input_df)
-        ft_id = ray.put(feature_transformers)
+        numpy_format_id = ray.put(numpy_format)
+        if not numpy_format:
+            input_df_id = ray.put(input_data)
```
Comment (yushan111) [inline pyzoo/zoo/automl/search/ray_tune_search_engine.py:322]: ray.put(input_data["df"])?

Comment (TheaperDeng) [reply pyzoo/zoo/automl/search/ray_tune_search_engine.py:322]: if it is not a numpy_format, input_data is a dataframe(see line140)

File: pyzoo/zoo/automl/search/ray_tune_search_engine.py [hunk 2]
```diff
@@ -89,6 +86,36 @@ def compile(self,
...
+        # data mode detection
+        assert isinstance(data, dict), 'ERROR: Argument \'data\' should be a dictionary.'
+        data_mode = None  # data_mode can only be 'dataframe' or 'ndarray'
+        data_schema = set(data.keys())
+        if set(["df"]).issubset(data_schema):
+            data_mode = 'dataframe'
+        if set(["x", "y"]).issubset(data_schema):
+            data_mode = 'ndarray'
+        assert data_mode in ['dataframe', 'ndarray'],\
+            'ERROR: Argument \'data\' should fit either dataframe schema or ndarray schema.'
```
Comment (yushan111) [inline pyzoo/zoo/automl/search/ray_tune_search_engine.py:98]: elaborate dataframe schema and ndarray schema.

File: pyzoo/zoo/automl/search/ray_tune_search_engine.py [hunk 3]
```diff
@@ -89,6 +86,36 @@ def compile(self,
...
+
+        # data extract
+        if data_mode == 'dataframe':
+            input_df = data['df']
+            feature_cols = data['feature_cols'] if 'feature_cols' in data.keys() else None
```
Comment (yushan111) [inline pyzoo/zoo/automl/search/ray_tune_search_engine.py:103]: data.get("feature_cols", None)

File: pyzoo/zoo/automl/search/ray_tune_search_engine.py [hunk 4]
```diff
@@ -89,6 +86,36 @@ def compile(self,
...
+            target_col = data['target_col'] if 'target_col' in data.keys() else None
+            validation_df = data['val_df'] if 'val_df' in data.keys() else None
+        else:
+            if data["x"].ndim == 1:
+                data["x"] = data["x"].reshape(-1, 1)
```
Comment (yushan111) [inline pyzoo/zoo/automl/search/ray_tune_search_engine.py:108]: input_data["x"] =

File: pyzoo/test/zoo/automl/searcher/test_ray_tune_search_engine.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            y = a*x + b
+            return x, y
+        train_x, train_y = get_linear_data(2, 5, 1000)
+        val_x, val_y = get_linear_data(2, 5, 400)
+        return train_x, train_y, val_x, val_y
+    
+    def create_date_dataset(self):
+        sample_num = np.random.randint(100, 200)
+        train_df = pd.DataFrame({"datetime": pd.date_range(
+            '1/1/2019', periods=sample_num), "value": np.random.randn(sample_num).astype(np.float32)})
```
[the remark this answers is not in the document]
Comment (TheaperDeng) [reply pyzoo/test/zoo/automl/searcher/test_ray_tune_search_engine.py:79]: it works fine with the change in `base_pytorch_model.py` line 55-59. I will remove the `.astype(np.float32)` here.

File: pyzoo/test/zoo/automl/searcher/test_ray_tune_search_engine.py [hunk 6]
```diff
@@ -40,22 +40,63 @@ def search_space(self, all_available_features):
...
+    search_space = recipe.search_space(feature_transformer.get_feature_list())\
+        if feature_transformer else None
+    searcher.compile(data=data,
+                    model_create_func=modelBuilder,
+                    recipe=recipe,
+                    feature_transformers=feature_transformer,
+                    search_space=search_space)
+    return searcher
+
+def get_data():
```
Comment (yushan111) [inline pyzoo/test/zoo/automl/searcher/test_ray_tune_search_engine.py:81]: get_numpy_data

File: pyzoo/test/zoo/automl/searcher/test_ray_tune_search_engine.py [hunk 7]
```diff
@@ -40,22 +40,63 @@ def search_space(self, all_available_features):
...
+    def get_linear_data(a, b, size):
+        x = np.arange(0, 10, 10 / size)
+        y = a*x + b
+        return x, y
+    train_x, train_y = get_linear_data(2, 5, 1000)
+    val_x, val_y = get_linear_data(2, 5, 400)
+    return train_x, train_y, val_x, val_y
+
+def create_date_dataset():
```
Comment (yushan111) [inline pyzoo/test/zoo/automl/searcher/test_ray_tune_search_engine.py:90]: get_ts_input

**Review by yushan111 — COMMENTED** [review]
Others LGTM

**Comment by TheaperDeng** [comment]
PR validation (combine) pass @yushan111 :
http://10.239.47.210:18888/job/ZOO-PR-Validation/4970/
http://10.239.47.210:18888/job/ZOO-PR-Python-Spark-2.4-py37-Horovod/853/
http://10.239.47.210:18888/job/ZOO-PR-Scala-Spark-2.4/3346/

**Comment by yushan111** [comment]
I will merge this PR and resolve further review comments in new PR.

**Pull request merged**